cmd/generator/main.go

```diff
@@ -213,39 +235,73 @@ func getReleasePlugin(ctx context.Context, client *github.Client, repositoryName
 			return nil, errors.Wrapf(err, "failed to read gzipped plugin bundle for release %s", releaseName)
 		}
 
-		bundleReader := tar.NewReader(gzBundleReader)
-		for {
-			hdr, err := bundleReader.Next()
-			if err == io.EOF {
-				break
-			}
-			if err != nil {
-				return nil, errors.Wrapf(err, "failed to read plugin bundle for release %s", releaseName)
-			}
+		bundleData, err := ioutil.ReadAll(gzBundleReader)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to read plugin bundle for release %s", releaseName)
+		}
+
+		manifestData, err := getTarFile(tar.NewReader(bytes.NewReader(bundleData)), "plugin.json")
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to read manifest from plugin bundle for release %s", releaseName)
+		}
+		plugin.Manifest = mattermostModel.ManifestFromJson(bytes.NewReader(manifestData))
 
-			if path.Base(hdr.Name) != "plugin.json" {
-				continue
+		if plugin.Manifest.IconPath != "" {
+			iconData, err := getTarFile(tar.NewReader(bytes.NewReader(bundleData)), plugin.Manifest.IconPath)
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to read icon data from plugin bundle for release %s", releaseName)
 			}
-			manifest := mattermostModel.ManifestFromJson(bundleReader)
 
-			plugin.Manifest = manifest
-			break
+			logger.Debugf("using icon specified in manifest as %s", plugin.Manifest.IconPath)
+			plugin.IconData = fmt.Sprintf("data:image/svg+xml;base64,%s", base64.StdEncoding.EncodeToString(iconData))
 		}
 	} else {
 		logger.Debugf("skipping download since found existing plugin")
 	}
 
+	if plugin.Manifest == nil {
+		return nil, fmt.Errorf("failed to find plugin manifest for release %s", releaseName)
+	}
+
 	// Reset fields, even if we found the existing plugin above.
-	plugin.HomepageURL = repository.GetHTMLURL()
-	plugin.IconData = ""
+	if plugin.Manifest.HomepageURL != "" {
+		plugin.HomepageURL = plugin.Manifest.HomepageURL
+	} else {
+		plugin.HomepageURL = repository.GetHTMLURL()
+	}
 	plugin.DownloadURL = downloadURL
 	plugin.DownloadSignature = []byte{}
+	plugin.UpdatedAt = updatedAt
 
-	if plugin.Manifest == nil {
-		return nil, fmt.Errorf("failed to find plugin manifest for release %s", releaseName)
+	return plugin, nil
+}
+
+func getTarFile(reader *tar.Reader, filepath string) ([]byte, error) {
+	for {
+		hdr, err := reader.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to read tar file")
+		}
+
+		// Match the filepath, assuming the tar file contains a leading folder matching the
+		// plugin id.
+		if matched, err := path.Match(fmt.Sprintf("*/%s", filepath), hdr.Name); err != nil {
```

Comment:
Maybe we want to be explicit here and check for the plugin id? 1/5

Reply:
Interestingly, at the moment I can't know the plugin id until I read `plugins.json`, which requires this generic match anyway.

Reply:
Well, that's a blocker. NVM then